### DISCUSSION TITLE
security question regarding server side instance authentication as admin

### DISCUSSION THREAD

**Discussion by realJogicodes** [pocketbase/pocketbase#4388] [Q&A] [answered]
I wonder if what I am doing is secure. You'll see what I mean. I handle most client `pb` interactions in my svelte kit app on the front end and use API rules to restrict access. Server-side stuff was tricky passing cookies all the time, but this way works quite well and I think is superior. Anyway, for context, I use pocketbase on the front end like this:

```
// $lib/pocketbase.ts
import PocketBase from 'pocketbase';
import { PUBLIC_PB_URL } from '$env/static/public';
export const pb = new PocketBase(PUBLIC_PB_URL);
```

And then I use that on the client in different components for various things:
```
<!-- any component that needs pb -->
<script>
  import {pb} from '$lib/pocketbase'
   // ... do stuff with it
</script>
```

That works. Now, for payments via stripe, I need to run server side code because the secret key can't be exposed. I've made a payments collection, which is admin only where I store details about payments such as price data, user id, payment success etc. To create an entry, I need to be admin and i run that in a `+page.server.ts` load function so that it never reaches the client:

```
// +page.server.ts
// server side code for /buy component

import { pb } from '$lib/pocketbase';
import { PRIVATE_PB_ADMIN_PWD } from '$env/static/private';
import { PRIVATE_PB_ADMIN_EMAIL } from '$env/static/

export async function load({ url }) {
  // ... more code omitted like stripe instantiation, payment data creation, irrelevant to question

  // store payment initiation in pocketbase
  const data = {
    // all the payment data
  }
  // authenticate admin
  const adminAuthRecord = await pb.admins.authWithPassword(PRIVATE_PB_ADMIN_EMAIL, PRIVATE_PB_ADMIN_PWD);
  // store data
  const record = await pb.collection('payments').create(data);
  // clear authentication❓❓❓
  // is this 👇 appropriate to make sure front end has not got admin privileges? Is it even needed?
  pb.authStore.clear()
  return {
    // stripe data for front end
  }
}
```

Is `pb.authStore.clear()` appropriate to remove the admin access on the backend after the record has been added? I could not find a `pb.admins...` method for that. (Is this needed in the first place?) This `pb` instance is not the same as the one used on my front end, since it runs on a different computer, even though I import it from the same file so my thinking is a bit muddled here. I don't think I need it but maybe, (I don't wanna create an exploit here and mess this up), so better to ask

Cheers!

**Comment by ganigeorgiev** [comment, marked as answer]
You don't need to call `pb.authStore.clear()` for the server-side admin client.

I guess the above is fine as long as you don't have mixed SvelteKit components (aka. components that can run both in the browser and server due to prerendering or something else).

There are 2 things to note:
- `authWithPassword()` is inherently slow due to the bcrypt hash verification and if the above is executed frequently it may be better to cache the auth state and reauthenticate only when the previous state has expired
- because your node server should be able to process requests from different users concurrently, having a single admin SDK instance will trigger the [Auto cancellation](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#auto-cancellation) and you may want to disable it when using the SDK as admin.

An alternative version could be to create a separate admin only client (eg. `$lib/server/admin.pocketbase.ts` or add `.server` to the filename) that will take care for the above, something like:

```js
import PocketBase from 'pocketbase'
import { PUBLIC_PB_URL } from '$env/static/public';
import { PRIVATE_PB_ADMIN_EMAIL, PRIVATE_PB_ADMIN_PWD } from '$env/static/private';

const pb = new PocketBase(PUBLIC_PB_URL);

pb.autoCancellation(false);

// the credentials can be filled via env variables
await pb.admins.authWithPassword(PRIVATE_PB_ADMIN_EMAIL, PRIVATE_PB_ADMIN_PWD, {
  // This will trigger auto refresh or auto reauthentication in case
  // the token has expired or is going to expire in the next 30 minutes.
  autoRefreshThreshold: 30 * 60
})

export default pb;
```

To use it you can just import it in your server actions:

```js
import adminClient from '$lib/server/admin.pocketbase.ts'

...

const record = await adminClient.collection('payments').create(data);
```

**Reply by realJogicodes** [reply]
Excellent, I think separating it out is smart, just in case. Maybe I do something dumb down the line, if admin and client pb instances are separated, that's better for resiliency. Thanks for that suggestion
